src/libsyntax/feature_gate.rs

```diff
@@ -259,16 +271,47 @@ const KNOWN_FEATURES: &'static [(&'static str, &'static str, Option<u32>, Status
 ];
 // (changing above list without updating src/doc/reference.md makes @cmr sad)
 
-enum Status {
-    /// Represents an active feature that is currently being implemented or
-    /// currently being considered for addition/removal.
-    Active,
+#[derive(PartialEq, Copy, Clone, Debug)]
+pub enum AttributeType {
+    /// Normal, builtin attribute that is consumed
+    /// by the compiler before the unused_attribute check
+    Normal,
 
-    /// Represents a feature which has since been removed (it was once Active)
-    Removed,
+    /// Builtin attribute that may not be consumed by the compiler
+    /// before the unused_attribute check. These attributes
+    /// will be ignored by the unused_attribute lint
+    Whitelisted,
 
-    /// This language feature has since been Accepted (it was once Active)
-    Accepted,
+    /// Builtin attribute that is only allowed at the crate level
+    CrateLevel,
+}
+
+pub enum AttributeGate {
+    /// Is gated by a given feature gate, reason
+    /// and function to check if enabled
+    Gated(&'static str, &'static str, fn(&Features) -> bool),
+
+    /// Ungated attribute, can be used on all release channels
+    Ungated,
+}
+
+// fn() is not Debug
+impl ::std::fmt::Debug for AttributeGate {
+    fn fmt(&self, fmt: &mut ::std::fmt::Formatter) -> ::std::fmt::Result {
+        match *self {
+            Gated(ref name, ref expl, _) => write!(fmt, "Gated({}, {})", name, expl),
+            Ungated => write!(fmt, "Ungated")
+        }
+    }
+}
+
+macro_rules! cfg_fn {
+    ($field: ident) => {{
+        fn f(features: &Features) -> bool {
+            features.$field
+        }
```

Comment:
Would it be possible to make `Features` an array rather than a struct and index into it using a constant for each feature, rather than this combination of function and macro? Or do we sometimes use a function other than the macro-generated one?